lib/instrumentation/modules/@aws-sdk/client-s3.js

```diff
@@ -152,17 +153,18 @@ function s3MiddlewareFactory (client, agent) {
 
           // Destination context.
           const destContext = {
-            address: context[elasticAPMStash].hostname,
-            port: context[elasticAPMStash].port,
             service: {
               name: SUBTYPE,
               type: TYPE
             }
           }
+          if (context[elasticAPMStash]) {
+            destContext.address = context[elasticAPMStash].hostname
+            destContext.port = context[elasticAPMStash].port
+          }
```

Comment:
Reviewer note: This is the fix for the second surprise. It is possible for middleware (if it doesn't call `next()`) to result in our "finalizeRequest" middleware not being called, so `context[elasticAPMStack]` wouldn't be set.